src/xshell_client.hpp

```diff
@@ -0,0 +1,64 @@
+/***************************************************************************
+* Copyright (c) 2016, Johan Mabille, Sylvain Corlay, Martin Renou          *
+* Copyright (c) 2016, QuantStack                                           *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+****************************************************************************/
+
+#ifndef XEUS_SHELL_CLIENT_HPP
+#define XEUS_SHELL_CLIENT_HPP
+
+#include "zmq.hpp"
+#include "nlohmann/json.hpp"
+#include "xeus/xkernel_configuration.hpp"
+#include "xeus-zmq/xauthentication.hpp"
+
+namespace xeus
+{
+
+    class xshell_client
+    {
+    public:
+
+        xshell_client(zmq::context_t& context,
+                        const std::string& user_name,
+                        const xeus::xconfiguration& config);
+
+        ~xshell_client();
+
+    protected:
+
+        void send_on_shell(nl::json header,
+                        nl::json parent_header,
+                        nl::json metadata,
+                        nl::json content);
+
+        nl::json receive_on_shell();
+
+        nl::json check_received_message(long timeout);
+
+        zmq::socket& get_socket();
```

Comment:
All these methods should be public, they'll be called from `xclient_zmq_impl`.

Reply:
Also, no need to specify `on_shell` in the methods' names here, since they are in the `xshell_client` class.